etcdmain/config.go

```diff
@@ -317,6 +356,141 @@ func (cfg *config) Parse(arguments []string) error {
 		}
 	}
 
+	return nil
+}
+
+func (cfg *config) configFromFile() error {
+	// Read in the contents of the config file.
+	b, err := ioutil.ReadFile(cfg.configFile)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal the YAML.
+	err = yaml.Unmarshal(b, cfg)
+	if err != nil {
+		return err
+	}
+
+	// Make sure the defaults are set for the missing fields.
+	cfg.setDefaultsIfNeeded()
+
+	// Validate whether conflicting parameters are specified.
+	nSet := 0
```

Comment:
This reminds of the question that I had about sharing the validations which was my initial intent.

For `configFromCmdLine` case we rely on `flags.IsSet` function to figure if the flag was actually specified. So if we move it to `Parse` function I am not sure how to get a similar functionality for the `configFromFile` case. As the fields are set to default values (by virtue of FlagSet init) we can’t simply look at the field values to determine if they were set (doing so some of the tests failed - `TestConfigParsingMissedAdvertiseClientURLsFlag #4`). Any inputs on how to go about this?

In that light I am not sure if the validations that I have put in place for `configFromFile` are appropriate.